### PR TITLE
Add an attribute type for jsonb with merge.

### LINKED ISSUE #65: Using the merge operator to update the jsonb column
Hey @jrochkind, 

I saw [your comment](https://github.com/rails/arel/issues/288#issuecomment-376032169) about adding support for the merge operator. I've been thinking about this problem myself. I started using your library recently and I've quickly realized the problem [you stated](https://github.com/jrochkind/attr_json#note-on-optimistic-locking) in the readme. Using optimistic locks everywhere I use attribute json columns isn't practical. We need a different solution. 

I've scoped out how Rails compiles the update SQL, and I think my solution is 90% there. Here is my rough plan for getting there:

- Monkey patch Arel::UpdateManager
```ruby
module Arel
  class UpdateManager < Arel::TreeManager
    def set values
      if String === values
        @ast.values = [values]
      else
        @ast.values = values.map { |column, value|
          klass = if column.is_jsonb?
                    Nodes::JsonMerge
                  else
                    Nodes::Assignment
                  end

          klass.new(
            Nodes::UnqualifiedColumn.new(column),
            value
          )
        }
      end
      self
    end
  end
end
```

- add `#is_jsonb?` method to Jsonb column type.
- create an `Arel::Nodes::JsonMerge` node
```ruby
module Arel
  module Nodes
    %w{
      JsonMerge
    }.each do |name|
      const_set name, Class.new(Binary)
    end
  end
end
```
- monkeypatch Arel::Visitors::BindVisitor to add a visitor
```ruby
module Arel
  module Visitors
    module BindVisitor
      def visit_Arel_Nodes_JsonMerge o, collector
        if o.right.is_a? Arel::Nodes::BindParam
          collector = visit o.left, collector
          collector << " || "
          visit o.right, collector
        else
          super
        end
      end
    end
  end
end
```

or

```ruby
module Arel
  module Visitors
    module PgJson
      private

      def visit_Arel_Nodes_JsonbMerge o, a
        json_infix o, a, '||'
      end

      def json_infix(o, a, opr)
        visit(Nodes::InfixOperation.new(opr, o.left, o.right), a)
      end
    end

    PostgreSQL.send :include, PgJson
  end
end
```

- Monkeypatch ActiveRecord::Persistence::ClassMethods#_update_record to replace the value of the attr_json column with the changes hash.

```ruby
module ActiveRecord::Persistence::ClassMethods
  def _update_record(values, constraints) # :nodoc:
    constraints = _substitute_values(constraints).map { |attr, bind| attr.eq(bind) }

    um = arel_table.where(
      constraints.reduce(&:and)
    ).compile_update(_substitute_values(_replace_jsonb_column(values)), primary_key)

    connection.update(um, "#{self} Update")
  end

  def _replace_jsonb_column(values)
    return values unless respond_to?(:attr_json_config)
    # Look up the name of the attr_json column
    container_name = attr_json_config.default_container_attribute
    # Determine if the key is in values hash
    return values unless values.key?(container_name)
    # Get the changes to this column
    change_arr = a.changes[container_name]
    changes = change_arr.last.reject {|k,v| change_arr.first[k] == v }
    # Replace the changes with the actual value in the `values` hash
    values[container_name] = changes
    values
  end
end
```

What do you think of this plan? It's not very clean, but it might work. Let me know if you'd be interested in a pull request or if it should be in its own repo.

### PR DESCRIPTION
**Pull request by volkanunsal**
Tasks

- [x] Integrate with `ContainerAttribute`
- [x] Write tests

Closes #65

### CODE REVIEW

**Comment by volkanunsal** [comment]
@jrochkind Let me know if you can think of any ways to test this....

**Comment by jrochkind** [comment]
Thanks! I'll try to find time this week to look at this. If you don't hear from me, feel free to ping me again.

**Comment by volkanunsal** [comment]
@jrochkind Ping!

**Comment by jrochkind** [comment]
thanks, hope to be able to look at this this next week. I anticipate wanting to play around with making some changes to it, do you have any preferences as to whether I do that in a new branch, or right in this branch?

**Comment by volkanunsal** [comment]
Great! I don't have any preference on the branch question.

**Comment by jrochkind** [comment]
Hmm, taking a look at this, I see it's still monkey-patching AR, and is *not* providing a new ActiveModel::Type?

Based on our previous discussion, I thought you had something that was based on a new ActiveModel::Type with limited or no AR monkey-patching?

I was really excited by that, which was the direction I was interested in, and am disappointed to not find it. :(  What's up with that direction @volkanunsal ?

File: lib/attr_json/type/container_attribute.rb [hunk 1]
```diff
@@ -1,14 +1,16 @@
...
       ActiveRecord::Type::Json
     else
       ActiveRecord::Type::Internal::AbstractJson
     end)
+
+    # A type that gets applied to the AR container/store jsonb attribute,
+    # to do serialization/deserialization/cast using declared attr_jsons, to
+    # json-able values, before calling super to original json-type, which will
+    # actually serialize/deserialize the json.
+    class ContainerAttribute < AttrJson::Type::SuperKlass
```
Comment (jrochkind) [inline lib/attr_json/type/container_attribute.rb]: I don't understand what this is doing. `SuperKlass` is either `ActiveRecord::Type::Json` or `ActiveRecord::Type::Internal::AbstractJson` (above). 

So what is `AttrJson::Type::SuperKlass`?  Does that turn into `AttrJson::Type::ActiveRecord::Internal::AbstractJson` or `AttrJson::Type::ActiveRecord::Type::Json`? I don't think so -- becuase those don't exist!

I'm wondering if it actually ends up being nothing but the same as `SuperKlass`, and no actualy change to previous code before this diff?

What were you actually trying to do here?

**Comment by jrochkind** [comment]
I'm talking about the code here: https://github.com/jrochkind/attr_json/issues/65#issuecomment-500217447

Which seemed to have a new type, the `AttrJson::Type::JsonbMerger` type, with only one small monkey-patch maybe only one method? (Which I'm actually a bit worried about, it seems to be monkey-patching ALL AR behavior, even when not using the new type, which seems like it would break standard AR behavior? But that's a different question). 

That code there does not seem to be what we're looking at here though? What happened to that approach?

**Comment by volkanunsal** [comment]
Yeah I was excited too and I really wanted to make it work but the problem with that approach was that serialize is used by not just UPDATE but also DELETE. So it would be too invasive. Plus the serialize wouldn’t have possession of the dirty attributes, which are only available in the record instance.

File: lib/attr_json/type/container_attribute.rb (hunk 1 above)
Comment (volkanunsal) [reply lib/attr_json/type/container_attribute.rb]: Oh this is from when I was trying to make the custom type work. I didn’t actually change the super class logic. So we can revert this part.

**Comment by volkanunsal** [comment]
This patches only update behavior. There is another method for delete and one other for create.

File: lib/attr_json/type/merger.rb [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+      end
+
+      # `substitude_values` builds query attribute and binds it to values.
+      # It uses the SQL literal above if the patch is enabled for the
+      # column type.
+      def self.substitude_values(klass, values)
+        values.map do |name, value|
+          attr = klass.arel_attribute(name)
+          old_bind = klass.predicate_builder.build_bind_attribute(name, value)
+          bind = enabled?(klass, name) ? literal(old_bind, value) : old_bind
```
Comment (volkanunsal) [inline lib/attr_json/type/merger.rb]: Regarding your point about "patching all AR behavior," this is the most important line to notice. Here, we check if the merge is enabled for the attribute, and only if it is we use the merge string. If not, we use the `old_bind` values. So this patch leaves other attributes completely unaffected.

File: lib/attr_json/type/merger.rb [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+          [attr, bind]
+        end
+      end
+
+      # `replace` replaces the value of the ContainerAttribute type with the
+      # changes to the container column and the removed keys. These will be used
+      # to build the SQL literal.
+      def self.replace(klass, values, changes)
+        values.map do |name, value|
+          if enabled?(klass, name)
```
Comment (volkanunsal) [inline lib/attr_json/type/merger.rb]: Same thing here. We check if the patch is enabled for the attribute first. Otherwise, it's unobtrusive.

**Comment by jrochkind** [comment]
OK, I see, thanks. 

So, the good news is that you're still only overriding *one* method in ActiveRecord, `_update_row`. This is good, as we want to minimize the number of overrides, ideally to one (and I _think_ better than a previous attempt that overrode a lot more). 

The bad news is, it's a pretty complex override. The comments say "This is the only part of the code that differs from the original code" , I think you mean the part that does something *different*, because the original method was actually only four (different) lines. I *think* it's [this one](https://github.com/rails/rails/blob/v5.2.3/activerecord/lib/active_record/persistence.rb#L695-L700). 

The nice thing about the previous implementation is that it was a very simple "If special_condition, then do special things; then call super anyway", which serves to make us more confident that for things that are *not* the special condition we won't accidentally break anything. 

With this code, I'm a bit worried that it might break other things we aren't thinking of. For instance, it looks to me like the implementation of AR optimistic locking *also* [overrides _update_row](https://github.com/rails/rails/blob/v5.2.3/activerecord/lib/active_record/locking/optimistic.rb#L79-L106).  By the override here without calling super, does it break optimistic locking? I think it probably does. That's just one I happened upon while looking around for original `_update_row` implementation; there could be other things we accidentally break. (Note that the optimistic override also has the desirable implementation pattern "Unless special thing, just call super"). 

And the use of so many AR internals in the implementation makes me worry that future changes in Rails would be likely to break things. 

It's possible this can be refactored to be a bit cleaner/safer. To be able to do that with confidence, we'd want good tests. 

One problem with the existing tests is -- the entire test file passes in _current attr_json master_ without any changes. So the tests might be good for making sure the changes didn't *break* anything (although they can't cover all of AR func to make sure we didn't break any of it) -- but they don't actually test that the new functionality is working. 

It's a bit tricky to figure out how to write tests that actually test the new func is working. What I can think of is either figuring out how to get at the actual SQL string generated by a `save` and test agains that; or set up somewhat tricky concurrency conditions to make sure that under the conditions this func is meant to change, changes don't overwrite each other. The first, testing actual SQL, is probably more useful (if we can figure out how to do it), although a combination could be done.  But one way or another, for a test to actually verify the new func, you need a test that *fails* without the new func, and succeeds with the new func. 

I admire you figuring this all out and getting something that works at all, these internals of AR are pretty hairy!  But unless we can figure out a way to make the code simpler and less intrusive, I won't personally feel confident in it, and probably won't want it in attr_json. It may be that AR just isn't structured in a way to make this kind of intervention possible reliably. (But if you want to put it in a third-party gem, I will try to figure out how to make that possible, so attr_json can still use it). We don't have to give up yet though (although I can't promise how much time I'll have to put into it); if you want to go further, I still think the first step is figuring out some proper tests.

**Comment by jrochkind** [comment]
For tests, if there's no way to get to the SQL, I suppose another way to test higher-level "integration" would be like:

1. Fetch the record with AR
2. Issuing a *direct* update to the record in the DB (perhaps using connection.update or whatever) make  a change to one key in the JSON. 
3. Make a key to another key in the fetched AR model and save. 
4. Reload the record from the DB. It should have both changes. Prior to merge code, the change in 2 will have been overwritten. 

I think that should work, and tests can be written for a variety of cases using that pattern -- changing an existing key, deleting a key, adding a new key, etc.

**Comment by jrochkind** [comment]
For other implementation ideas, looking at the original implementation:

```ruby
    def _update_row(attribute_names, attempted_action = "update")
      self.class._update_record(
        attributes_with_values(attribute_names),
        self.class.primary_key => id_in_database
      )
    end
```

I wonder if we can simply pass a different thing in to the first argument of `_update_record`, attributes with values. Like what if the *value* for the JSONb attribute we pass in becomes the string "existing_column_name || {new: hash}::jsonb"?  

The answer is probably that some other part of AR will mess it up trying to 'escape' it. But it may escape values differently for different underlying types, so it may be possible to affect *that* with a custom Type? 

But these are just brainstorming thoughts without really engaging with the code., this approach might not work at all (and you might already know it won't from your existing investigations!).   Having tests that only pass when the feature is working will provide a harness to experiment with the code and see if different approaches can work.

**Comment by volkanunsal** [comment]
Thanks for the feedback. I'll take a closer look at your points soon. As far as your point about `_update_record`, this method is actually folded into `_update_row` in my implementation. You see, `_update_record` is used by other methods besides `_update_row`, but we only need it to be patched when it's called from `_update_row`. That's why, I had to bring its internals up the call chain, where I could escape the query properly.

I think we can also take a page from Optimistic lock module and call super unless `enabled?` returns true for the current record.

**Comment by jrochkind** [comment]
Yeah, I understand _update_record is folded into _update_row.  I just don't like it, heh. It makes it sensitive to underlying changes to _update_row in future releases, since it copies and pastes and fixes in place the code. Among other things, it just introduces more possibilities of errors. But it could work. 

I am worried that there's actually no way to override _update_row *without* breaking optimistic locking, if any logic branches don't call super. If you don't call super, you won't get optimistic locking behavior.  And what other things might be broken we don't even know? But first things first I guess. 

I am worried that there may be no reliable way to do this without changes to AR. But perhaps we can get to the point of finding a fairly simple change to AR we can PR to rails for future versions. 

Trustworthy tests are definitely the first step, if you can get some trustworthy tests, I will try to find time to hack on it a bit (although it might not be right away).

**Comment by jrochkind** [comment]
I asked on Stack Overflow if anyone figured out a way to get the SQL that would be generated by a `save`, but haven't received any great answers. https://stackoverflow.com/questions/56860315/can-i-see-what-sql-an-activerecord-save-call-would-generate-without-doing-the/

**Comment by volkanunsal** [comment]
> if anyone figured out a way to get the SQL that would be generated by a save

The way I would do it is to create a wrapper for `update` that generates the sql from `arel_table`.

```ruby
> scope = SqlWrapper.new(Project.where(name: 'bar')).update(name: 'foo')
> scope.to_sql
# => "UPDATE projects SET name = 'foo' WHERE name = 'bar';"
> scope.run
```

And I would use the SQL generation logic below to implement `.to_sql`. Here it the relevant test from Arel test suite:

```ruby
table   = Table.new :users
manager = Arel::SelectManager.new
manager.from table
manager.take 1
stmt = manager.compile_update(Nodes::SqlLiteral.new('foo = bar'), Arel::Attributes::Attribute.new(table, 'id'))
stmt.key = table['id']

stmt.to_sql.must_be_like %{
  UPDATE "users" SET foo = bar
  WHERE "users"."id" IN (SELECT "users"."id" FROM "users" LIMIT 1)
}
```

**Comment by jrochkind** [comment]
`SqlWrapper.new(Project.where(name: 'bar')).update(name: 'foo')`

Isn't the same thing as `.save` though. I'm not sure how it helps us write tests for "verify we get the right SQL when we call `save`", which is what this patch does?  You can try to write an update that does the "same thing" but you're counting on getting it right, which lessons one's trust in the tests. (Plus that kind of udpate doesn't involve dirty attributes at all, I think it skips the main part of what the jsonb update merge stuff *does*!)

I'm not finding that approach useful for testing this stuff, but maybe I'm not seeing it right.

**Comment by volkanunsal** [comment]
As far as verifying the SQL, what I'm thinking of is to plug in this wrapper after `compile_update` in `_update_row`, and have it send a `ActiveSupport::Notification` with the generated SQL when it's being tested. Then the test can subscribe to the notification and validate the SQL.

**Comment by jrochkind** [comment]
I wonder if just doing the real "integration" test is easier and higher confidence anyway.  But your call, you're writing it!

But test the actual behavior we are trying to do, rather than the SQL that accomplishes it. 

```ruby
model = Model.new(json_attributes: { "one" => "1", "two" => "2"}).save!

Model.where(id: model.id).update("two" => "new 2")

model.json_attributes["three"] = "3"
model.save!

model.reload
expect(model.json_attributes["two"]).to eq("new 2")
```

I think that should fail on existing master, and pass if and only if using json update merge functionality.

**Comment by volkanunsal** [comment]
Aha good idea!

**Comment by jrochkind** [comment]
So I think it's safe to say this ended up much more complicated to do in a general and reliable way than hoped for. 

Are you still working on it @volkanunsal ? Would you like us to leave the PR open, or close it for now?

**Comment by volkanunsal** [comment]
@jrochkind Thanks for ping. I've been meaning to come back to it for a while now. I keep getting pulled into do other things. It's definitely a useful feature for us, so I'd like to finish it. I have one more project to wrap up this week and then I'll come back and finish this up this weekend.

**Comment by jrochkind** [comment]
OK! I'm honestly leaning toward thinking I may not want to include it in attr_json. I think it's ending up more complicated and less universally reliable than I want to put in the attr_json codebase at present.  Like I probably don't want to include a feature in attr_json that works only if you are _not_ using nested json hashes, since a large part of the purpose of attr_json is to support nested json hashes!

But I'm happy to keep looking at it, possibly for the purpose of figuring out what, if any, changes we need to make in attr_json to allow you to do this as an add-on locally or in another gem. Or to try to help you simplify it, and/or to see if I can be happy including it in attr_json after all.

**Comment by jrochkind** [comment]
Also not happy with breaking optimistic locking. You might want to write a test for optimistic locking just to verify if you have or haven't broken it.

**Comment by volkanunsal** [comment]
I think what I want to do is to make this an opt-in feature. It could work the same way optimistic locking does and be enabled by a mixin.

**Comment by jrochkind** [comment]
AR Optimistic Locking is in a mix-in, but it's mixed into ActiveRecord:Base, the end-user doesn't actually get to choose whether to mix it in or not. Although one could imagine an AR refactor that worked like that. 

But also, mixing in optimistic locking does not _break_ any other AR features. I think that's teh biggest downside of how I think you've figured out how to do this so far -- it _breaks_ existing AR features (Optimistic Locking). 

But another possible outcome of this code is figuring out the smallest change that could be made to AR to allow this to be put in _without_ breaking Optimistic Locking... and then trying to PR that to Rails?

**Comment by volkanunsal** [comment]
I may not be able to get to this soon. I'm swamped. Sorry. You can leave this open or not. It's up to you.

**Comment by jrochkind** [comment]
Hi, I'm going to close this for now -- I'm sorry I put roadblocks in your way. This *would* be a very exciting feature. 

But it's is a tricky thing to do reliably that will work in all cases without breaking anything else, and since ActiveRecord doesn't have great API to support it and requires "monkey patching", it risks breaking in future Rails versions. 

I'm just reluctant to add more code complexity to attr_json of this sort, that would need to be maintained as bugs as reported and future versions of Rails are released. I think attr_json is already pushing the boundaries of being maintainably sustainably with mostly just me -- although it's still passing tests on Rails github master ("edge"), which is a good sign -- I want to keep it that way!

So sorry closing for now. I really appreciate your interest in attr_json, and your work demo'ing this -- the code will still be here on github for others to look at. 

I'd encourage you to look into doing this separately from attr_json, as a separate gem, that works with any AR json column. If we need to make changes to attr_json to allow it to work with such a separate gem -- I can do that!  I'm thinking we might just need to allow someone to configure what "parent" ActiveModel::Type to use for the container column, so it could be configured to use one from such a third-party gem. I can do that if needed!

**Pull request closed without merging**